### PR TITLE
models: implement simple fixedwing model

### PR DESCRIPTION
**Pull request by sjschlapbach**
This pull request introduces a modified model for a fixedwing standard plane with linear lift model (function of angle of attack and elevator input) and quadratic drag model.

An additional "--normalization" flag was introduced to have the ability to disable normalization when using simulation data or the new flight log standards with setpoints. The model can be run with the same makefile command as mentioned in the README:
```
make estimate-model model=fixedwing_longitudinal_model log=resources/elevator_ramp_zero_thrust.csv
```

For VTOL configurations (with different PX4 setpoint topics for the correct control surfaces and thrust quantities), use the following configuration:
```
make estimate-model model=vtol_longitudinal_model log=LOG_PATH
```

### CODE REVIEW

File: Tools/parametric_model/src/models/dynamics_model.py [hunk 1]
```diff
@@ -609,161 +628,184 @@ def plot_scatter(ax, title, dataframe_x, dataframe_y, dataframe_z, color='blue')
...
             # fisher_information_m_mat[i]= np.trace(fisher_information_matrix_m)
 
-            ## condition number
+            # condition number
             # fisher_information_f_mat[i]= min(np.abs(np.linalg.eigvals(fisher_information_matrix_f))) / \
             # max(np.abs(np.linalg.eigvals(fisher_information_matrix_f)))
             # fisher_information_m_mat[i]= min(np.abs(np.linalg.eigvals(fisher_information_matrix_m))) / \
-            # max(np.abs(np.linalg.eigvals(fisher_information_matrix_m)))
\ No newline at end of file
+            # max(np.abs(np.linalg.eigvals(fisher_information_matrix_m)))
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/src/models/dynamics_model.py:811]: The changes in this file seems to be format fixes. Could we make this a separate PR so it would be easier to find out what changed?

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: Tools/parametric_model/src/models/simple_fixedwing_model.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+
+import numpy as np
+
+from .dynamics_model import DynamicsModel
+from .model_config import ModelConfig
+from .aerodynamic_models import LinearWingModel, ControlSurfaceModel
+
+
+class FixedWingPlaneModel(DynamicsModel):
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/src/models/simple_fixedwing_model.py:47]: name the file as the class name

File: Tools/parametric_model/src/models/dynamics_model.py (hunk 1 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/models/dynamics_model.py]: Removed formatting changes - the file was not changed at all, formatting was done by an auto-formatter

File: Tools/parametric_model/src/models/simple_fixedwing_model.py (hunk 2 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/models/simple_fixedwing_model.py]: done

**Comment by Jaeyoung-Lim** [comment]
@sjschlapbach I think I am getting a different issue as you have proposed
```
-------------------------------------------------------------------------------
Complete results saved to: 
model_results/simple_fixedwing_model_2022-12-01-15-48-51.yaml
-------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jaeyoung/.local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2898, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 70, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 101, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 1675, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 1683, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'measured_force_x'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "Tools/parametric_model/generate_parametric_model.py", line 131, in <module>
    start_model_estimation(**vars(arg_list))
  File "Tools/parametric_model/generate_parametric_model.py", line 114, in start_model_estimation
    model.plot_model_predicitons()
  File "/home/jaeyoung/dev/data-driven-dynamics/Tools/parametric_model/src/models/dynamics_model.py", line 572, in plot_model_predicitons
    "measured_force_x", "measured_force_y", "measured_force_z", 'blue')
  File "/home/jaeyoung/dev/data-driven-dynamics/Tools/parametric_model/src/models/dynamics_model.py", line 537, in plot_scatter
    ax.scatter(self.data_df[dataframe_x], self.data_df[dataframe_y],
  File "/home/jaeyoung/.local/lib/python3.6/site-packages/pandas/core/frame.py", line 2906, in __getitem__
    indexer = self.columns.get_loc(key)
  File "/home/jaeyoung/.local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2900, in get_loc
    raise KeyError(key) from err
KeyError: 'measured_force_x'
Makefile:34: recipe for target 'estimate-model' failed
make: *** [estimate-model] Error 1
```

**Comment by sjschlapbach** [comment]
> @sjschlapbach I think I am getting a different issue as you have showed in the meeting.
> 
> Is this consistent of what you are expecting in this branch?
> 
> ```
> -------------------------------------------------------------------------------
> Complete results saved to: 
> model_results/simple_fixedwing_model_2022-12-01-15-48-51.yaml
> -------------------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/jaeyoung/.local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2898, in get_loc
>     return self._engine.get_loc(casted_key)
>   File "pandas/_libs/index.pyx", line 70, in pandas._libs.index.IndexEngine.get_loc
>   File "pandas/_libs/index.pyx", line 101, in pandas._libs.index.IndexEngine.get_loc
>   File "pandas/_libs/hashtable_class_helper.pxi", line 1675, in pandas._libs.hashtable.PyObjectHashTable.get_item
>   File "pandas/_libs/hashtable_class_helper.pxi", line 1683, in pandas._libs.hashtable.PyObjectHashTable.get_item
> KeyError: 'measured_force_x'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "Tools/parametric_model/generate_parametric_model.py", line 131, in <module>
>     start_model_estimation(**vars(arg_list))
>   File "Tools/parametric_model/generate_parametric_model.py", line 114, in start_model_estimation
>     model.plot_model_predicitons()
>   File "/home/jaeyoung/dev/data-driven-dynamics/Tools/parametric_model/src/models/dynamics_model.py", line 572, in plot_model_predicitons
>     "measured_force_x", "measured_force_y", "measured_force_z", 'blue')
>   File "/home/jaeyoung/dev/data-driven-dynamics/Tools/parametric_model/src/models/dynamics_model.py", line 537, in plot_scatter
>     ax.scatter(self.data_df[dataframe_x], self.data_df[dataframe_y],
>   File "/home/jaeyoung/.local/lib/python3.6/site-packages/pandas/core/frame.py", line 2906, in __getitem__
>     indexer = self.columns.get_loc(key)
>   File "/home/jaeyoung/.local/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 2900, in get_loc
>     raise KeyError(key) from err
> KeyError: 'measured_force_x'
> Makefile:34: recipe for target 'estimate-model' failed
> make: *** [estimate-model] Error 1
> ```
> 
> To reproduce:
> 
> ```
> make estimate-model model=fixedwing_longitudinal_model log=resources/elevator_ramp_zero_thrust.csv 
> ```

Plotting currently doesn't work for this model, but the estimation can still be run with the following command:
```
python3 Tools/parametric_model/generate_parametric_model.py --plot False --config Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml resources/elevator_ramp_zero_thrust.csv
```

File: Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+import math
+import numpy as np
+
+from scipy.spatial.transform import Rotation
+from progress.bar import Bar
+
+
+class LinearWingModel():
+    def __init__(self, config_dict, mass):
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py:47]: I don't think the linear wing model needs the mass information of the vehicle

File: Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        #                       angular_acceleration[2]) / (self.air_density * self.ref_area * (v_airspeed[0]**2 + v_airspeed[2]**2))
+        # X_wing_aero_frame[0, 3] = - dyn_pressure
+        # X_wing_aero_frame[0, 4] = - (1 / (np.pi * np.e)) * dyn_pressure * cL**2
+
+        # features for lift coefficient computation
+        X_wing_aero_frame[2, 0] = - dyn_pressure * self.ref_area
+        X_wing_aero_frame[2, 1] = - dyn_pressure * \
+            self.ref_area * angle_of_attack
+        X_wing_aero_frame[2, 2] = - dyn_pressure * \
+            self.ref_area * elevator_input
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py:91]: Here you are only computing features the first three coefficients: 'c_L_0', 'c_L_alpha', `c_L_delta` and the features for `c_D_0`, `inv_ar` are always zero. Is this expected? If this is the case, we would always have `c_d_0` and `inv_ar` estimating zero

File: Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py (hunk 3 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py:47]: If the drag coefficient estimation should be done in the same way as before (in the `longitudinal-identification` repository), the mass is required to extract the measured lift coefficient from the available quantities in the dataset (see commented code in lines 80ff). Alternatively, it would only be possible to reparametrize the identification and to model drag as an e.g. quadratic function of the angle of attack.

File: Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py (hunk 4 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py:91]: Yes, this was done on purpose to first get the lift parameter identification up and running and then implement the drag estimation in a next step. I wouldn't be aware of any issues that might occur, despite possibly senseless results for `c_D_0` and `inv_ar`, which we can ignore for the lift estimation.

Comment (Jaeyoung-Lim) [reply Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py:91]: @sjschlapbach I did a small experiment of removing those from the problem and the estimation results are influenced. So I think there is something related to this

**Comment by Jaeyoung-Lim** [comment]
Before I review the PR, would it be possible to address the CI failures?
e.g.
```
Starting data resampling of topic types:  dict_keys(['actuator_outputs', 'vehicle_local_position', 'vehicle_attitude', 'vehicle_angular_velocity', 'sensor_combined'])
Traceback (most recent call last):
  File "Tools/parametric_model/generate_parametric_model.py", line 133, in <module>
    start_model_estimation(**vars(arg_list))
  File "Tools/parametric_model/generate_parametric_model.py", line 74, in start_model_estimation
    model = getattr(models, model_class)(config, normalization=normalization)
TypeError: __init__() got an unexpected keyword argument 'normalization'
make: *** [Makefile:34: estimate-model] Error 1
```

File: Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py (hunk 4 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py]: I'm gonna resolve this as the entire implementation was changed

File: Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py (hunk 3 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/models/aerodynamic_models/linear_wing_model.py]: yes, has been removed right now with some code clean up

**Comment by Jaeyoung-Lim** [comment]
@sjschlapbach Could we try to separate the changes that we want to merge in and what belongs outside?

Then we can try to merge the changes in

File: Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml [hunk 5]
```diff
@@ -0,0 +1,164 @@
+# general information
+model_name: "Gazebo Standard Plane"
+model_type: "Standard Plane Longitudinal Model"
+model_class:
+  "SimpleFixedWingModel"
+
+  # all vectors in FRD body frame if not specified otherwise
+model_config:
+  mass: 20.0
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml:9]: Why is the vehicle so heavy? Is this because this is coming from some paper?

File: Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  moment_of_inertia:
+    Ixx: 25.0
+    Iyy: 20.0
+    Izz: 20.0
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml:13]: Compared to the mass, the moment of inertia seems quite high. Does this come from some reference model? If so, can you cite the source?

File: Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+          description: "aileron_left"
+          dataframe_name: "u5"
+
+        - control_surface_2:
+          description: "elevator"
+          dataframe_name: "u7"
+
+        - control_surface_3:
+          description: "rudder"
+          dataframe_name: "u2"
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml:48]: Why does a longitudinal flight dynamics model have ailerons and rudders?

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]

File: Tools/parametric_model/generate_parametric_model.py [hunk 8]
```diff
@@ -127,5 +127,7 @@ def start_model_estimation(config, log_path, data_selection="none", plot=False):
                         help='Configuration file path for pipeline configurations')
     parser.add_argument('--plot', metavar='plot', type=str2bool, default='True',
                         help='Show plots after fit.')
+    parser.add_argument('--normalization', metavar='normalization', type=str2bool, default='True', required=False,
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/generate_parametric_model.py:130]: I think the default should be false, since this would break all the other models

Comment (sjschlapbach) [reply Tools/parametric_model/generate_parametric_model.py:130]: If I understood it correctly, normalization was always applied whenever the topic "V_air_body_x" was missing from the data frame (line 87ff in /Tools/parametric_model/src/models/dynamics_model.py). Maybe we could remove this however, if we don't want to run simulation files through this pipeline or add the missing topics already to the csv-file before.

File: Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml (hunk 7 above)
Comment (sjschlapbach) [reply Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml:48]: Currently, they are just part of this model in order not to generate errors during the parsing. I'm however still planning on removing this dependency, so that they can be left out for this model at least.

File: Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml (hunk 5 above)
Comment (sjschlapbach) [reply Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml:9]: Honestly, I think I just inserted some random values as they only scale the resulting parameters anyways and should not influence tracking. I'm going to look into values that make more sense.

File: Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml (hunk 6 above)
Comment (sjschlapbach) [reply Tools/parametric_model/configs/fixedwing_longitudinal_model.yaml:13]: Same as for the mass, these values were basically randomly selected, but should only result in a scaling of the result as well. Setting them all to 1.0 and using the identified coefficient led to exactly the same fit as before. But they should definitely be adapted to a value that is more consistent with the test log.

File: Tools/parametric_model/src/models/simple_fixedwing_model.py [hunk 9]
```diff
@@ -87,8 +87,8 @@ def prepare_force_regression_matrices(self):
         throttle = self.data_df["throttle"].to_numpy()
 
         # vector with cl0, clalpha, cldelta, cd0, cdalpha, cdalpha_sq, ct
-        coeff_vec = np.array([1.0911507971564856, 6.47680410045709, 2.2670575265057487,
-                             0.11373832931995426, 0.7433772294557139, 2.8798589221670072, 85.7567143535296])
+        coeff_vec = np.array([0.07910843279384513, 0.46956829728313765, 0.1643616706716667,
+                             0.00824602887569668, 0.053894849135539054, 0.20878977185710684, 6.217361790630906])
```
Comment (Jaeyoung-Lim) [inline Tools/parametric_model/src/models/simple_fixedwing_model.py:91]: Why does this need to be hard coded here?

Comment (sjschlapbach) [reply Tools/parametric_model/src/models/simple_fixedwing_model.py:91]: no specific reason - the fitted coefficients are not yet available at this point in the pipeline and all this plotting is for debugging only anyways. I'll remove all of it for sure before we merge it (there are corresponding comments in the beginning and end of the two sections)

Comment (Jaeyoung-Lim) [reply Tools/parametric_model/src/models/simple_fixedwing_model.py:91]: The framework already provides predictions with estimated parameters right? There is also a separate app where it can do prediction without estimation

Comment (sjschlapbach) [reply Tools/parametric_model/src/models/simple_fixedwing_model.py:91]: It does, but I think I broke the plotting at some point for the new model and didn't fix it yet. I didn't know about the separate app though. I will get it done until next week at the latest.

File: Tools/parametric_model/src/tools/dataframe_tools.py [hunk 10]
```diff
@@ -32,35 +32,102 @@
...
 
-    if pwm_threshold is None:
-        pwm_threshold = PWM_THRESHOLD
+        print('\nProcesed {0} thrust ramps and {1} pitch ramps'.format(
+            num_thrust_ramps, num_pitch_ramps))
 
-    if control_threshold is None:
-        control_threshold = ACTUATOR_CONTROLS_THRESHOLD
-    act_df_crp = act_df[act_df.iloc[:, 2] > pwm_threshold]
-    act_df_crp = act_df[act_df.iloc[:, 4] > pwm_threshold]
```
Comment (sjschlapbach) [inline Tools/parametric_model/src/tools/dataframe_tools.py:59]: If I understand it correctly, the line before this one won't have an effect as the result is overwritten anyways? The new implementation that checks if all actuator outputs are above the pwm threshold (if no selection parameter is used) is probably too restrictive on the other way to detect if the aircraft is airborne... What was the idea behind this flight detection?

File: Tools/parametric_model/generate_parametric_model.py (hunk 8 above)
Comment (sjschlapbach) [reply Tools/parametric_model/generate_parametric_model.py:130]: Normalization can be removed as soon as all models (and possibly the logs as well) have been adapted to include vehicle_torque_setpoints and thrust_setpoints (already normalized). Normalization is only required to obtain these values from the actuator outputs.

File: Tools/parametric_model/src/tools/dataframe_tools.py (hunk 10 above)
Comment (sjschlapbach) [reply Tools/parametric_model/src/tools/dataframe_tools.py:59]: The flight time computation has been completely reworked now and is based on the 'landed' prop of the vehicle status topic. This should be more reliable and generally applicable to many different types of aircraft.

**Review by Jaeyoung-Lim — CHANGES_REQUESTED** [review]
Thanks! 
Could you resolve the conflicts?

**Review by Jaeyoung-Lim — COMMENTED** [review]
@sjschlapbach How was the quadrotor_model data generated? They seem to have been modifield

**Comment by sjschlapbach** [comment]
> @sjschlapbach How was the quadrotor_model data generated? They seem to have been modifield

@Jaeyoung-Lim The new flight time compation (in `dataframe_tools`) is based on the `vehicle_status/landed` topic, which was missing from the csv files (as it was not used before). To ensure that the estimation still works with these logs, I added the `landed` column and filled it with zeros.

**Comment by Jaeyoung-Lim** [comment]
@sjschlapbach Thanks! 

Bombs away!
